lib/fs.js

```diff
@@ -1590,12 +1590,68 @@ fs.realpath = function realpath(path, options, callback) {
   if (!nullCheck(path, callback))
     return;
   var req = new FSReqWrap();
-  req.oncomplete = callback;
+  req.oncomplete =
+      (err, resolvedPath) =>
+        interceptELOOP(err, resolvedPath, path, options, callback);
   binding.realpath(pathModule._makeLong(path), options.encoding, req);
   return;
 };
 
 
+function interceptELOOP(err, resolvedPath, path, options, callback) {
+  if (resolvedPath)
+    return callback(err, resolvedPath);
+  if (err && err.code !== 'ELOOP')
+    return callback(err);
+  var retPath = '';
+  // Start at -1 since first operation is to + 1.
+  var offset = -1;
+  var current = 0;
+  var slashCount = 0;
+  var callDepth = 0;
+
+  // Can assume '/' because uv_fs_realpath() cannot return UV_ELOOP on win.
+  (function pathResolver(err2, resolvedPath2) {
+    // callDepth is too much. Return original error.
+    if (++callDepth > 100) return callback(err);
+    // No need to handle an error that was returned by a recursive call.
+    if (err2) return callback(err2);
+    // Done iterating over the path.
+    if (offset === path.length) return callback(null, resolvedPath2);
+
+    retPath = resolvedPath2;
+    slashCount = countSlashes(retPath);
+    offset = get32SlashOffset(path, offset + 1, slashCount);
+
+    var tmpPath = retPath + path.slice(current, offset);
+    current = offset;
+    var req = new FSReqWrap();
+    req.oncomplete = pathResolver;
+    binding.realpath(pathModule._makeLong(tmpPath), options.encoding, req);
+  }(null, ''));
+}
```

Comment:
I agree. Reason I did it this way was because handling/passing the async resources was being a thorn at the time. I'll see about moving this into C++.